src/set_theory/principal.lean

```diff
@@ -0,0 +1,66 @@
+/-
+Copyright (c) 2022 Violeta Hernández Palacios. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Violeta Hernández Palacios
+-/
+
+import set_theory.ordinal_arithmetic
+
+/-!
+### Principal ordinals
+
+We define principal or indecomposable ordinals, and we prove the standard properties about them.
+
+### Todo
+* Prove the characterization of additive indecomposable ordinals.
+* Prove the characterization of multiplicative indecomposable ordinals.
+* Refactor any related theorems from `ordinal_arithmetic` into this file.
+-/
+
+universe u
+
+namespace ordinal
+
+/-! ### Principal ordinals -/
+
+/-- An ordinal `o` is said to be principal or indecomposable under an operation when the set of
+ordinals less than it is closed under that operation. In standard mathematical usage, this term is
+almost exclusively used for additive and multiplicative principal ordinals.
+
+For simplicity, we break usual convention and regard 0 as principal. -/
+def principal (op : ordinal → ordinal → ordinal) (o : ordinal) : Prop :=
+∀ ⦃a b⦄, a < o → b < o → op a b < o
```

Comment:
This predicate looks like it's preserved under swapping the argument order to `op`, maybe it's useful to have that as a lemma, and then you get things like `iterate_lt_of_principal` for the other argument too. In particular since addition and multiplication aren't commutative for you!

Reply:
Added the lemma `principal_iff_principal_swap` and `iterate_lt'` for good measure.

Reply:
Actually, on second thought, I'd rather not add `iterate_lt'` unless I need it later down the line. I can't currently see a need for lemmas like these, and I'd rather not add every minor variation on a lemma unless it's actually useful.

Reply:
Yeah I think omitting `iterate_lt'` is totally fine since it should just be an easy consequence of `principal_iff_principal_swap` and `iterate_lt`